src/core/manager.rs

```diff
@@ -924,4 +997,30 @@ mod tests {
         assert_eq!(wm.focused_client().unwrap().id(), 20);
         assert_eq!(clients(&mut wm), vec![40, 30, 50, 10, 20]);
     }
+
+    #[test]
+    fn getting_all_clients_on_workspace() {
+        let conn = MockXConn::new(test_screens(), vec![]);
+        let mut wm = wm_with_mock_conn(test_layouts(), &conn);
+
+        add_n_clients(&mut wm, 3, 0); 
+        wm.focus_workspace(&Selector::Index(1));
+        add_n_clients(&mut wm, 2, 3); 
+
+        assert_eq!(wm.all_clients(&Selector::Condition(&|c: &Client| c.workspace() == 0)).len(), 3);
+        assert_eq!(wm.all_clients_mut(&Selector::Condition(&|c: &Client| c.workspace() == 1)).len(), 2);
+    }
+
+    #[test]
+    fn getting_all_workspaces_of_window() {
```

Comment:
Perhaps this test will make more sense when or if we can have a window on multiple workspaces ("tags" system).

Reply:
Ah, this is going to be a little fragile I think: you should be able to set things up so that a `Client` is on multiple workspaces (though you might need to use `push_workspace` currently to do so). If you do that with say 5 or 6 workspaces and run the tests multiple times I suspect you'll see intermittent failures: the `iter` method on `HashMap` doesn't guarantee ordering. We could sort the result before returning I suppose but that would require implementing `PartialOrd` for each `T` we care about (should be doable with derive in most cases) and then moving the underlying methods on `Ring` to being constrained by `T: PartialOrd`

Reply:
Okay, I see what you mean. I'll think about it, maybe I can do something.

Reply:
I'm a bit confused at the moment. I don't see where we are getting an iterator from a `HashMap`. `Ring` uses `VecDeque`, so I think it is already ordered, but I could be wrong. I'm guessing you're talking about the iterator over `client_map` in `all_clients`. That does not affect `all_workspaces` and its test though. Maybe it would be useful if `all_clients` would return an ordered vector if that's what you mean, but I don't know how that relates to `Ring`.

I also don't know how you would add a `Client` to multiple workspaces, because you can only have one `Client` in `client_map` for each `WinId` and each `Client` can currently only be on one workspace. I know you talked about using hooks to implement this in #32, but I don't understand how it would go together with `all_workspaces`, because currently `all_workspaces` can at most return a single workspace when given a `WinId` selector.

Reply:
ah, sorry! Yeah this comment is in relation to `all_clients`  (the mention of tagging is what sent me down that train of thought). `all_workspaces` is fine but this _is_ going to be an issue for `all_clients`: the client map is 1-1 but the iteration order is random, so if you do something like pass a condition that selects all `Clients` that represent a given program (say chrome) then you will get those clients in a random order. We probably want to do something like order them by WinId as that would (i think) be equivalent to ordering them by creation time. (I'd need to double check that holds but ordering by WinId would be fine regardless)

Reply:
I understand, now it all makes sense! I'll try to implement something tomorrow.